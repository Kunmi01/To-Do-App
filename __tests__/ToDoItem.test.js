import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import ConnectedToDoItem, {
  ToDoItem
} from '../src/components/ToDoList/components/ToDoItem';
import {
  updateToDoItem,
  deleteToDoItem,
  createToDoItem
} from '../src/redux/actions';

describe('ToDoItem', () => {
  const initialState = {};
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it('renders connected(SMART) component', () => {
    const container = mount(
      <Provider store={store}>
        <ConnectedToDoItem />
      </Provider>
    );
    expect(container.find(ConnectedToDoItem).exists()).toBe(true);
    expect(toJson(container)).toMatchSnapshot();
  });

  it('can CREATE a todo', () => {
    const container = mount(
      <Provider store={store}>
        <ConnectedToDoItem creation editable />
      </Provider>
    );
    const toDoItemContainer = container.find(ToDoItem);

    renderer.act(() => {
      toDoItemContainer.find('.todo-item__form__name').simulate('change', {
        target: { value: 'test name' }
      });
    });
    renderer.act(() => {
      toDoItemContainer
        .find('.todo-item__form__description')
        .simulate('change', {
          target: { value: 'test description' }
        });
    });
    renderer.act(() => {
      toDoItemContainer.find('.todo-item__form').simulate('submit');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      createToDoItem({
        itemId: expect.any(String),
        itemName: 'test name',
        itemDescription: 'test description',
        itemCreatedDate: new Date().toString()
      })
    );
  });

  it('can UPDATE a todo', () => {
    const container = mount(
      <Provider store={store}>
        <ConnectedToDoItem
          dispatch={store.dispatch}
          id="1"
          name="test name"
          description="test description"
          createdDate="today"
        />
      </Provider>
    );
    const toDoItemContainer = container.find(ToDoItem);

    renderer.act(() => {
      toDoItemContainer
        .find('.todo-item__form__button--edit')
        .simulate('click');
    });
    renderer.act(() => {
      toDoItemContainer
        .find('.todo-item__form__description')
        .simulate('change', {
          target: { value: 'description changed' }
        });
    });
    renderer.act(() => {
      toDoItemContainer.find('.todo-item__form').simulate('submit');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      updateToDoItem({
        itemId: '1',
        itemName: 'test name',
        itemDescription: 'description changed',
        itemCreatedDate: 'today'
      })
    );
  });

  it('can DELETE a todo', () => {
    const container = mount(
      <Provider store={store}>
        <ConnectedToDoItem
          dispatch={store.dispatch}
          id="1"
          name="test name"
          createdDate={new Date().toString()}
        />
      </Provider>
    );
    const toDoItemContainer = container.find(ToDoItem);

    renderer.act(() => {
      toDoItemContainer
        .find('.todo-item__form__button--delete')
        .simulate('click');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      deleteToDoItem({ itemId: '1' })
    );
  });
});
