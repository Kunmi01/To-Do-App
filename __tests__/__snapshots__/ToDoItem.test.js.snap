// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem renders connected(SMART) component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ToDoItem)>
    <ToDoItem
      createdDate=""
      creation={false}
      description=""
      dispatch={[MockFunction]}
      editable={false}
      id=""
      name=""
    >
      <li
        className="todo-item"
      >
        <form
          className="todo-item__form"
          onSubmit={[Function]}
        >
          <input
            className="todo-item__form__name"
            disabled={true}
            id="name"
            onChange={[Function]}
            placeholder="Add a to-do..."
            type="text"
            value=""
          />
          <div
            className="todo-item__form__bottom-wrapper"
          >
            <p
              className="todo-item__form__created"
            >
              Created: 
            </p>
            <div
              className="todo-item__form__buttons-wrapper"
            >
              <button
                className="todo-item__form__button todo-item__form__button--edit"
                onClick={[Function]}
                type="button"
              >
                Edit
              </button>
              <button
                className="todo-item__form__button todo-item__form__button--delete"
                onClick={[Function]}
                type="button"
              >
                Delete
              </button>
            </div>
          </div>
        </form>
      </li>
    </ToDoItem>
  </Connect(ToDoItem)>
</Provider>
`;
